namesys/routing.go

```diff
@@ -59,6 +59,18 @@ func (r *IpnsResolver) resolveOnceAsync(ctx context.Context, name string, option
 	}
 
 	name = strings.TrimPrefix(name, "/ipns/")
+
+	// Fix PeerID represented as CIDv1 with invalid multicodec
+	if name[:4] == "bafy" { // TODO: is this a safe way of detecting base32-dag-pb?
+		ipnsCid, err := cid.Decode(name)
+		if err == nil && ipnsCid.Type() != cid.Libp2pKey {
+			// CIDs in IPNS are expected to have libp2p-key multicodec.
+			// We ease the transition by fixing multicodec on the fly:
+			// https://github.com/ipfs/go-ipfs/issues/5287#issuecomment-492163929
+			name = cid.NewCidV1(cid.Libp2pKey, ipnsCid.Hash()).String()
```

Comment:
Shouldn't we return a redirect?

Reply:
We do return redirect on the gateway.

This is a fix for CLI where redirect is not possible.
User pasting `/ipns/{cidv1-dag-pb}` could produce error because `peer` returns error if multicodec in CIDv1 is other than `libp2p`.

Reply:
Ah. I'm wondering if we should just return an error that includes the correct path? I'm concerned that if it "just works" with no visible redirect, it'll effectively become a feature.

Reply:
That is a very good point. I agree returning error with already fixed CID is a fair compromise outside of gateway.

How does this look?

```console
$ ipfs resolve /ipns/bafybeidsocs4bhczzvd7ilegoppt23cfdydexjkomn3dm245zzeqziz5ge
Error: peer ID represented as CIDv1 require libp2p-key multicodec: retry with bafzbeidsocs4bhczzvd7ilegoppt23cfdydexjkomn3dm245zzeqziz5ge
```

Reply:
Maybe use `/ipns/bafzbe`...?

Reply:
Good idea, done in ba682045a110b9bea2274af88689c594e378a090